slideflow/io/preservedsite/crossfolds.py

```diff
@@ -2,12 +2,129 @@
 import numpy as np
 import pandas as pd
 
-import slideflow as sf
-from slideflow.util import log
+# import slideflow as sf
+# from slideflow.util import log
+from itertools import combinations
+import random
+
+
+def generate_test_data():
+
+    k = 3
+    patients = [f'pt{p}' for p in range(200)]
+    sites = [f'site{s}' for s in range(5)]
+    outcomes = list(range(4))
+    category = 'outcome_label'
+    df = pd.DataFrame({
+        'patient': pd.Series([random.choice(patients) for _ in range(100)]),
+        'site': pd.Series([random.choice(sites) for _ in range(100)]),
+        'outcome_label': pd.Series([random.choice(outcomes) for _ in range(100)])
+    })
+    unique_labels = df['outcome_label'].unique()
+    return [df, category, unique_labels, k]
+
+def generate_brute_force(data, category, values, crossfolds=3, target_column='CV3',
+             patient_column='patient', site_column='site',
+             timelimit=10):
+    # New dataframe for data
+    submitters = data[patient_column].unique()
+    newData = pd.merge(
+        pd.DataFrame(submitters, columns=[patient_column]),
+        data[[patient_column, category, site_column]],
+        on=patient_column, how='left'
+    )
+    newData.drop_duplicates(inplace=True)
+    unique_sites = data[site_column].unique()
+
+    # list of possible combinations of sites in folds; also built in check for use case when someone chooses the wrong number of folds
+    if crossfolds > len(unique_sites):
+        print("choose less number of crossfolds")
+    else:
+        most_possible_sites_in_one_fold = 1 + int(len(unique_sites)) - crossfolds
+        list_of_all_combos = list()
+        for i in range(most_possible_sites_in_one_fold):
+            list_of_all_combos += list(combinations(unique_sites, i+1))
+        list_of_proper_combos = list(combinations(list_of_all_combos, crossfolds))
+        removal_list = list()
+        for item in list_of_proper_combos:
+            item_length = 0
+            list_of_items_in_a_combo = list()
+            for item2 in item:
+                item_length += len(item2)
+                list_of_items_in_a_combo.extend(item2)

```

Comment:
Since the item is a list of lists and I need the list of sites from all of those lists; I changed it to a list comprehension instead.

sites_in_a_possible_crossfold = [site for site in possible_crossfold] where
- possible_crossfold was item
- where sites_in_a_possible_crossfold was list_of_items_in_a_combo
@jamesdolezal does that work better

Reply:
This line:

`sites_in_a_possible_crossfold = [site for site in possible_crossfold] `

is equivalent to:

`sites_in_a_possible_crossfold = possible_crossfold` 

If this is what you are intending, you should consider removing the `sites_in_a_possible_crossfold` entirely and just use `possible_crossfold`. Does that make sense, or am I missing something?